File: tests/check_site.py
#!/usr/bin/env python3
"""Static guardrails for the Zoe Life concept. No third-party packages."""
from __future__ import annotations

import html.parser
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HTML = ROOT / "index.html"
CSS = ROOT / "css" / "style.css"
JS = ROOT / "js" / "main.js"
SCENE = ROOT / "js" / "scene.js"
FILES = [HTML, CSS, JS, SCENE]
results: list[tuple[bool, str, str]] = []


def check(ok: bool, name: str, detail: str) -> None:
    results.append((ok, name, detail))


class Collector(html.parser.HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self.tags: list[tuple[str, dict[str, str], int]] = []
        self.ids: set[str] = set()
        self.duplicate_ids: set[str] = set()
        self.headings: list[int] = []
        self.forms: list[dict[str, str]] = []
        self.controls: list[tuple[str, dict[str, str], int]] = []
        self.labels_for: set[str] = set()

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        data = {key: value or "" for key, value in attrs}
        line = self.getpos()[0]
        self.tags.append((tag, data, line))
        if "id" in data:
            if data["id"] in self.ids:
                self.duplicate_ids.add(data["id"])
            self.ids.add(data["id"])
        if re.fullmatch(r"h[1-6]", tag):
            self.headings.append(int(tag[1]))
        if tag == "form":
            self.forms.append(data)
        if tag in {"input", "select", "textarea"}:
            self.controls.append((tag, data, line))
        if tag == "label" and data.get("for"):
            self.labels_for.add(data["for"])


missing = [str(path.relative_to(ROOT)) for path in FILES if not path.is_file()]
check(not missing, "Core files exist", ", ".join(missing) if missing else "HTML, CSS and JavaScript found")
if missing:
    for ok, name, detail in results:
        print(("PASS" if ok else "FAIL"), name, detail)
    sys.exit(1)

html_text = HTML.read_text(encoding="utf-8")
css_text = CSS.read_text(encoding="utf-8")
js_text = JS.read_text(encoding="utf-8")
scene_text = SCENE.read_text(encoding="utf-8")
parser = Collector()
parser.feed(html_text)

check(html_text.lstrip().lower().startswith("<!doctype html>"), "Document essentials", "doctype present")
check(html_text.count("<main") == 1 and html_text.count("<h1") == 1, "Single main and h1", "one main and one h1")
check(not parser.duplicate_ids, "IDs are unique", ", ".join(sorted(parser.duplicate_ids)) if parser.duplicate_ids else "no duplicate IDs")

heading_skip = any(current > previous + 1 for previous, current in zip(parser.headings, parser.headings[1:]))
check(not heading_skip, "Heading order", "levels do not skip")

problems: list[str] = []
for tag, attrs, line in parser.tags:
    for attribute in ("src", "href"):
        value = attrs.get(attribute, "")
        if not value or value.startswith(("#", "data:", "mailto:", "tel:")):
            continue
        if re.match(r"^[a-z]+://", value):
            problems.append(f"line {line}: external {value}")
            continue
        target = (ROOT / value.split("?")[0].split("#")[0]).resolve()
        if not target.exists():
            problems.append(f"line {line}: missing {value}")
check(not problems, "Local references resolve", "; ".join(problems) if problems else "all linked files exist and no external URL is loaded")

unlabelled = []
for tag, attrs, line in parser.controls:
    if attrs.get("type") == "radio" and not attrs.get("id"):
        continue
    control_id = attrs.get("id")
    if not control_id or control_id not in parser.labels_for:
        unlabelled.append(f"{tag} line {line}")
check(not unlabelled, "Form controls are labelled", ", ".join(unlabelled) if unlabelled else "all named fields have labels")

form_safe = all("action" not in form and "method" not in form for form in parser.forms)
form_safe = form_safe and "fetch(" not in js_text and "XMLHttpRequest" not in js_text
form_safe = form_safe and "Nothing is sent, stored or emailed" in html_text
check(form_safe, "Mock form cannot submit", "no action, method or network request; visible disclosure present")

required_assets = [
    "assets/current-site/founders-current-site.jpg",
    "assets/current-site/about-guidance.jpg",
    "assets/current-site/coaching-zoe-life.jpg",
    "assets/current-site/services-professional.jpg",
    "assets/founder-tayo.jpg",
    "assets/founder-kemi.jpg",
]
missing_assets = [asset for asset in required_assets if asset not in html_text or not (ROOT / asset).is_file()]
check(not missing_assets, "Verified photos and covers are used", ", ".join(missing_assets) if missing_assets else "founders, current-site photography and both verified covers present")

manifest_path = ROOT / "assets" / "current-site" / "manifest.json"
manifest_problems = []
if not manifest_path.is_file():
    manifest_problems.append("manifest missing")
else:
    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    for asset in manifest.get("assets", []):
        if not (manifest_path.parent / asset["file"]).is_file():
            manifest_problems.append(asset["file"])
        if not asset.get("source_page", "").startswith("https://www.zoelifehub.com/"):
            manifest_problems.append(f"{asset['file']} source page")
check(not manifest_problems, "Current-site provenance is complete", ", ".join(manifest_problems) if manifest_problems else "manifested assets and source pages resolve")

season_missing = [name for name in ["building", "waiting", "growing", "healing", "leading"] if f'data-season="{name}"' not in html_text]
focus_missing = [name for name in ["spiritual", "relational", "personal", "professional"] if f'value="{name}"' not in html_text]
path_missing = [name for name in ["Relationships and Family", "Academic and Career", "Faith and Life Resources"] if name not in html_text]
check(not season_missing and not focus_missing and not path_missing, "Season and pathfinder content", f"missing: {season_missing + focus_missing + path_missing}" if season_missing or focus_missing or path_missing else "five seasons, four focuses and three pathways present")

forbidden_copy = [
    "Book a consult", "Book a 20-minute consult", "Get the book", "landing end of August",
    "Amazon", "Gumroad", "Selar", "On the roadmap", "In the works", "data-soon",
    "person-photo--t", "person-photo--k", "custom cursor",
]
found_copy = [term for term in forbidden_copy if term.lower() in html_text.lower()]
check(not found_copy, "Unverified offers and placeholders removed", ", ".join(found_copy) if found_copy else "no invented availability, marketplaces, roadmap claims or avatar placeholders")

page_prose = re.sub(r"<script\b[^>]*>.*?</script>", "", html_text, flags=re.S | re.I)
page_prose = re.sub(r"<style\b[^>]*>.*?</style>", "", page_prose, flags=re.S | re.I)
forbidden_dashes = re.findall(r"[—–]", page_prose)
check(not forbidden_dashes, "Visible copy has no forbidden dash characters", f"found {len(forbidden_dashes)}" if forbidden_dashes else "zero em dash and en dash characters")

external_patterns = []
for path, text in [(HTML, html_text), (CSS, css_text), (JS, js_text), (SCENE, scene_text)]:
    runtime_text = re.sub(r"data:image[^\"')]+", "", text, flags=re.I)
    if re.search(r"https?://|//(?:cdn|unpkg|cdnjs|fonts)\.", runtime_text, re.I):
        external_patterns.append(path.name)
check(not external_patterns, "No external dependencies", ", ".join(external_patterns) if external_patterns else "all runtime dependencies are local")

motion_ok = "prefers-reduced-motion" in css_text and "prefers-reduced-motion" in js_text and "prefers-reduced-motion" in scene_text
motion_ok = motion_ok and "window.addEventListener('scroll'" not in js_text and 'window.addEventListener("scroll"' not in js_text
check(motion_ok, "Motion is reduced and scroll-safe", "reduced motion in CSS and JS; no raw window scroll listener")

hidden_main = bool(re.search(r"\.js\s+\.reveal\s*\{[^}]*opacity:\s*0", css_text, re.S))
check(not hidden_main, "Main content fails open", "no hidden-by-default reveal system")

print("\nZoe Life checks")
print("=" * 92)
failures = 0
for ok, name, detail in results:
    if not ok:
        failures += 1
    print(f"{'PASS' if ok else 'FAIL':<5} {name:<42} {detail}")
print("=" * 92)
print(f"{len(results) - failures} passed, {failures} failed\n")
sys.exit(1 if failures else 0)

File: tests/check-site.mjs
/**
 * Zoe Life Phase 1 site checks. No dependencies.
 *   node tests/check-site.mjs
 *
 * Regression checks for the things most likely to go wrong on a client build:
 * broken links, missing labels, invented facts, exposed private addresses,
 * leftover stock, and forms that lie about succeeding.
 */

import { readFileSync, existsSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const PAGES = ["index.html", "about.html", "books.html", "connect.html", "contact.html", "consult.html"];
const REDIRECTS = ["family-life.html", "appointments.html"];

let pass = 0;
const failures = [];

function check(name, ok, detail) {
  if (ok) {
    pass++;
    console.log(`  ok   ${name}`);
  } else {
    failures.push(`${name}${detail ? ` :: ${detail}` : ""}`);
    console.log(`  FAIL ${name}${detail ? ` :: ${detail}` : ""}`);
  }
}

function group(title) {
  console.log(`\n${title}`);
}

const read = (f) => readFileSync(join(ROOT, f), "utf8");
const html = Object.fromEntries([...PAGES, ...REDIRECTS].map((p) => [p, read(p)]));
const js = read("js/main.js");
const css = read("css/style.css");

function visibleText(doc) {
  return doc
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<!--[\s\S]*?-->/g, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/&[a-z]+;/gi, " ")
    .replace(/\s+/g, " ");
}

function headerOf(doc) {
  const start = doc.indexOf('<header class="site-header">');
  const end = doc.indexOf("<main id=\"main\">");
  return start >= 0 && end > start ? doc.slice(start, end) : "";
}

function footerOf(doc) {
  const start = doc.indexOf('<footer class="site-footer">');
  return start >= 0 ? doc.slice(start) : "";
}

function mainOf(doc) {
  const start = doc.indexOf('<main id="main">');
  const end = doc.indexOf("</main>");
  return start >= 0 && end > start ? doc.slice(start, end) : "";
}

function flattenValues(obj) {
  const out = [];
  for (const v of Object.values(obj)) {
    if (v && typeof v === "object") out.push(...flattenValues(v));
    else out.push(v);
  }
  return out;
}

/* ----------------------------------------------------------- structure -- */
group("Structure and SEO");

const titles = new Map();
const descs = new Map();

for (const p of PAGES) {
  const doc = html[p];
  const title = (doc.match(/<title>([^<]*)<\/title>/) || [])[1] || "";
  const desc = (doc.match(/<meta name="description" content="([^"]*)"/) || [])[1] || "";

  check(`${p} has a title`, title.length > 10, title);
  check(`${p} has a meta description`, desc.length > 40, `${desc.length} chars`);
  titles.set(p, title);
  descs.set(p, desc);

  const h1s = doc.match(/<h1[\s>]/g) || [];
  check(`${p} has exactly one h1`, h1s.length === 1, `found ${h1s.length}`);

  check(`${p} has a main landmark`, /<main id="main">/.test(doc));
  check(`${p} has a skip link`, /class="skip-link"/.test(doc));
  check(`${p} has header and footer landmarks`, /<header class="site-header">/.test(doc) && /<footer class="site-footer">/.test(doc));
  check(`${p} declares lang`, /<html lang="en">/.test(doc));

  const levels = [...doc.matchAll(/<h([1-4])[\s>]/g)].map((m) => Number(m[1]));
  let skipped = null;
  for (let i = 1; i < levels.length; i++) {
    if (levels[i] > levels[i - 1] + 1) skipped = `h${levels[i - 1]} then h${levels[i]}`;
  }
  check(`${p} heading order has no skipped level`, !skipped, skipped || "");
}

check("All page titles are unique", new Set(titles.values()).size === PAGES.length);
check("All meta descriptions are unique", new Set(descs.values()).size === PAGES.length);

/* ------------------------------------------------ deployability -------- */
group("Deployability");

const SITE = "https://www.zoelifehub.com";
for (const p of PAGES) {
  const doc = html[p];
  const canon = (doc.match(/<link rel="canonical" href="([^"]+)"/) || [])[1];
  check(`${p} has a canonical URL`, !!canon && canon.startsWith(SITE), canon);
  check(`${p} declares og:url`, doc.includes(`<meta property="og:url"`));
}
const canons = PAGES.map((p) => (html[p].match(/rel="canonical" href="([^"]+)"/) || [])[1]);
check("Canonical URLs are unique per page", new Set(canons).size === PAGES.length);
check("Home canonical has no trailing filename", canons[0] === SITE + "/", canons[0]);

for (const f of ["404.html", "robots.txt", "sitemap.xml", "favicon.svg", "js/config.js"]) {
  check(`${f} is generated`, existsSync(join(ROOT, f)));
}

const robots = read("robots.txt");
const sitemap = read("sitemap.xml");
const isStaging = /noindex/.test(html["index.html"]);
if (isStaging) {
  check("Staging build disallows crawlers", /Disallow: \//.test(robots));
} else {
  check("Production build allows crawlers", /Allow: \//.test(robots));
  check("robots.txt points at the sitemap", robots.includes(`${SITE}/sitemap.xml`));
  check("Pages are indexable", PAGES.every((p) => /content="index, follow"/.test(html[p])));
}
check("Sitemap lists every page", PAGES.every((p) => sitemap.includes(canonicalOf(p))));
check("Sitemap omits moved-url stubs", !sitemap.includes("/family-life") && !sitemap.includes("/appointments"));
check("Sitemap is well formed", /^<\?xml/.test(sitemap) && /<\/urlset>/.test(sitemap));

const notFound = read("404.html");
check("404 page links home", /href="index\.html"/.test(notFound));
check("404 page has one h1", (notFound.match(/<h1[\s>]/g) || []).length === 1);

function canonicalOf(p) {
  return SITE + "/" + (p === "index.html" ? "" : p.replace(/\.html$/, ""));
}

for (const p of REDIRECTS) {
  check(`${p} exists as a moved-url page`, existsSync(join(ROOT, p)));
  check(`${p} has one h1`, (html[p].match(/<h1[\s>]/g) || []).length === 1);
}
check("family-life.html points to Connect", /href="connect\.html"/.test(html["family-life.html"]));
check("appointments.html points to Consult", /href="consult\.html"/.test(html["appointments.html"]));

/* --------------------------------------------------------------- links -- */
group("Links");

for (const p of [...PAGES, ...REDIRECTS]) {
  const doc = html[p];

  const internal = [...doc.matchAll(/href="(?!https?:|mailto:|#)([^"]+)"/g)].map((m) => m[1]);
  const broken = internal.filter((h) => !existsSync(join(ROOT, h.split("#")[0])));
  check(`${p} internal links all resolve`, broken.length === 0, broken.join(", "));

  const anchors = [...doc.matchAll(/href="#([^"]+)"/g)].map((m) => m[1]);
  const missingAnchor = anchors.filter((a) => !doc.includes(`id="${a}"`));
  check(`${p} in-page anchors all have targets`, missingAnchor.length === 0, missingAnchor.join(", "));

  const ext = [...doc.matchAll(/<a[^>]*href="https?:[^"]*"[^>]*>/g)].map((m) => m[0]);
  const unsafe = ext.filter((a) => !/rel="noopener noreferrer"/.test(a));
  check(`${p} external links use rel=noopener noreferrer`, unsafe.length === 0, `${unsafe.length} missing`);

  const fake = (doc.match(/example\.(com|org)|lorem|TODO|FIXME|href="#"/gi) || []);
  check(`${p} has no placeholder or fake URLs`, fake.length === 0, fake.join(", "));
}

/* -------------------------------------------------------------- images -- */
group("Images");

for (const p of PAGES) {
  const imgs = [...html[p].matchAll(/<img[^>]*>/g)].map((m) => m[0]);
  const noAlt = imgs.filter((i) => !/\salt=/.test(i));
  check(`${p} every image has an alt attribute`, noAlt.length === 0, noAlt.join(" | "));

  const noDims = imgs.filter((i) => !/width=/.test(i) || !/height=/.test(i));
  check(`${p} images declare width and height`, noDims.length === 0, `${noDims.length} missing`);

  const named = imgs.filter((i) => /tayo|kemi|akinyemi/i.test(i));
  for (const img of named) {
    const alt = (img.match(/\salt="([^"]*)"/) || [])[1] || "";
    check(`Named photo alt includes Tayo or Kemi: ${alt.slice(0, 60)}`, /Tayo|Kemi/i.test(alt), alt);
  }
}

const allAssets = new Set();
for (const p of [...PAGES, ...REDIRECTS]) {
  for (const m of html[p].matchAll(/(?:src|href)="(assets\/[^"]+)"/g)) allAssets.add(m[1]);
}
for (const m of css.matchAll(/url\(\.\.\/(assets\/[^)]+)\)/g)) allAssets.add(m[1]);
const missingAssets = [...allAssets].filter((a) => !existsSync(join(ROOT, a)));
check("All referenced assets exist", missingAssets.length === 0, missingAssets.join(", "));

const publishedHtml = PAGES.map((p) => html[p]).join("\n");
check("No Unsplash or Pexels assets", !/unsplash|pexels/i.test(publishedHtml + css));
check("No leftover stock filenames", !/hands-reaching|coaching-conversation|small-group-study|cream-plaster/i.test(publishedHtml));

/* --------------------------------------------------------------- forms -- */
group("Forms");

for (const p of PAGES) {
  const doc = html[p];
  const controls = [...doc.matchAll(/<(input|select|textarea)[^>]*>/g)].map((m) => m[0]);

  const noId = controls.filter((c) => !/\sid="/.test(c));
  check(`${p} every form control has an id`, noId.length === 0, `${noId.length} missing`);

  const ids = controls.map((c) => (c.match(/\sid="([^"]+)"/) || [])[1]).filter(Boolean);
  const unlabelled = ids.filter((id) => !doc.includes(`for="${id}"`));
  check(`${p} every form control has a label`, unlabelled.length === 0, unlabelled.join(", "));

  const required = controls.filter((c) => /\srequired/.test(c));
  const noDescribe = required.filter((c) => !/aria-describedby="/.test(c));
  check(`${p} required fields have aria-describedby`, noDescribe.length === 0, `${noDescribe.length} missing`);
}

const contact = html["contact.html"];
for (const opt of [
  "Coaching",
  "Speaking Engagement",
  "Workshop / Group Session",
  "Academic or Career Support",
  "Books &amp; Resources",
  "Collaboration / Partnership",
  "General Inquiry",
  "Other",
]) {
  check(`Contact reason option present: ${opt}`, contact.includes(`>${opt}<`) || contact.includes(`value="${opt}"`));
}
check("Contact form has a conditional Other field", /data-reveal="Other"/.test(contact));
check("Contact form has a honeypot", /name="website"/.test(contact));
check("Contact form includes first, last, email, phone, message",
  ["c-first", "c-last", "c-email", "c-phone", "c-message"].every((id) => contact.includes(`id="${id}"`)));
check("Contact form is separate from consultation booking",
  /data-form="contact"/.test(contact) && existsSync(join(ROOT, "consult.html")));
check("Subscribe forms are email-only",
  ![...publishedHtml.matchAll(/data-form="subscribe"[\s\S]*?<\/form>/g)].some((m) => /type="tel"|name="phone"/.test(m[0])));

/* ------------------------------------------------------------- honesty -- */
group("Honesty and safety");

check("Success is only rendered from renderSent", (js.match(/renderSent\(/g) || []).length >= 1);
check(
  "renderSent is called only inside a res.ok branch",
  /if \(res\.ok\) \{\s*renderSent\(/.test(js),
  "renderSent must be guarded by res.ok"
);
check(
  "Submission is attempted only when an endpoint is configured",
  /if \(!endpoint\) \{[\s\S]{0,200}renderBlocked/.test(js)
);
check("Blocked state is explicit", /not been delivered|not been sent|have not been subscribed/i.test(js));
check("Failure is never dressed as success", /renderFailed[\s\S]{0,400}did not send/i.test(js));

const cfg = read("js/config.js");
check("config.js exists and defines ZOE_CONFIG", /window\.ZOE_CONFIG\s*=/.test(cfg));
const cfgObj = JSON.parse(cfg.slice(cfg.indexOf("{"), cfg.lastIndexOf("}") + 1));
for (const k of ["formEndpoint", "newsletterEndpoint", "bookingUrl"]) {
  check(`config.${k} is present`, k in cfgObj);
}
check("config.payments is present", cfgObj.payments && typeof cfgObj.payments === "object");
check(
  "No placeholder endpoint was invented",
  flattenValues(cfgObj).every((v) => v === null || /^https:\/\//.test(String(v))),
  JSON.stringify(cfgObj)
);

check(
  "No Zoe Life email address is exposed",
  !/[a-z0-9._%+-]+@zoelifehub\.com/i.test(publishedHtml),
  (publishedHtml.match(/[a-z0-9._%+-]+@zoelifehub\.com/i) || [])[0]
);
check("No mailto links", !/mailto:/i.test(publishedHtml));
check("No private backend addresses", !/@yahoo\.com|@gmail\.com/i.test(publishedHtml));
check("No prices are stated", !/\$\s?\d|USD\s?\d|\d+\.\d{2}\s?(?:USD|dollars)/i.test(visibleText(publishedHtml)));
check("KingsWord is not listed as a client", !/kingsword/i.test(publishedHtml));

const booksDoc = html["books.html"];
check("No invented storefront URLs", !/amazon\.com|etsy\.com|gumroad\.com|selar\.co/i.test(booksDoc));
check("No leftover Link pending chips", !/Link pending/.test(booksDoc));
check("No Cover pending placeholder", !/Cover pending/.test(booksDoc));
check("Devotional cover is present", /assets\/books\/gratitude-devotional-cover\.jpg/.test(booksDoc));
check("Journal cover is present", /assets\/books\/gratitude-journal-cover\.jpg/.test(booksDoc));
check("Purchase options are fail-closed when unconfigured", /Purchase options coming/.test(booksDoc) || /Pay with Stripe/.test(booksDoc));
check("No couple workbook", !/Questions Every Christian Couple|questions-before-marriage/i.test(publishedHtml));
check("Books page has both Saturday titles",
  /A 7-Day Gratitude Devotional/.test(booksDoc) && /A 100-Day Gratitude Journal/.test(booksDoc));

const consult = html["consult.html"];
if (!cfgObj.bookingUrl) {
  check("Consult shows GOOGLE_CALENDAR_BOOKING_URL placeholder", /GOOGLE_CALENDAR_BOOKING_URL/.test(consult));
} else {
  check("Consult booking uses a live https URL", consult.includes(cfgObj.bookingUrl));
}
check(
  "No leftover Acuity sales copy",
  !/Ready for a Breakthrough|secure your slot|direct transformation/i.test(publishedHtml)
);

/* --------------------------------------------------------------- brand -- */
group("Brand structure");

const TAGLINE = "Helping people thrive in every season of life.";
check("Tagline appears on Home as a sentence", html["index.html"].includes(TAGLINE));
check("Tagline appears on About", html["about.html"].includes(TAGLINE));
check("Tagline is not inside the logo alt", !/alt="[^"]*Thrive[^"]*"/.test(publishedHtml));
const indexHeader = headerOf(html["index.html"]);
check("Header uses the banner wordmark", /class="brand-wordmark"[^>]*src="assets\/brand\/zoe-life-wordmark\.jpg"/.test(html["index.html"]));
check("Header does not use the circular mark as the logo", !indexHeader.includes("zoe-life-mark.png"));
check("Footer uses the circular mark", footerOf(html["index.html"]).includes("assets/brand/zoe-life-mark.png"));

check("Header has no cart", !/\(0\)|\bCart\b|\bLogin\b|\bAccount\b/i.test(indexHeader));
check("Primary CTA is the 20-minute consultation", /Schedule a complimentary 20-minute consultation/.test(indexHeader));

const navBlock = html["index.html"].match(/<nav class="site-nav"[\s\S]*?<\/nav>/)[0];
for (const label of ["Home", "About", "Books &amp; Resources", "Connect", "Contact"]) {
  check(`Main nav includes ${label}`, navBlock.includes(`>${label}<`));
}
check("Family Life is not a main nav label", !/>Family Life</.test(navBlock));

check("Home shows Pastors Tayo and Kemi", /tayo-kemi-full\.jpg/.test(html["index.html"]));
check("About shows Pastors Tayo and Kemi", /tayo-kemi-about\.jpg/.test(html["about.html"]));
check("Consult shows Pastors Tayo and Kemi", /tayo-kemi-full\.jpg/.test(consult));
check("Contact shows Pastor Kemi", /kemi-white\.jpg/.test(contact));

const ZOE_LIFE = ["facebook.com/zoelifehub", "instagram.com/zoelifehub", "tiktok.com/@zoelifehub1", "youtube.com/@zoelifehub1", "x.com/zoelifehub"];
const ZFL = ["facebook.com/zoefamilylife10", "instagram.com/zoefamilylife", "tiktok.com/@zoefamilylife", "youtube.com/@zoefamilylife"];

for (const p of PAGES) {
  const footer = footerOf(html[p]);
  for (const s of ZOE_LIFE) check(`${p} footer links Zoe Life ${s.split("/")[0]}`, footer.includes(s));
  const zflInFooter = ZFL.filter((s) => footer.includes(s));
  check(`${p} footer does NOT use Zoe Family Life accounts`, zflInFooter.length === 0, zflInFooter.join(", "));
}

const connectMain = mainOf(html["connect.html"]);
check("Connect introduces Zoe Life before Zoe Family Life",
  connectMain.indexOf("The main accounts") < connectMain.indexOf("Zoe Family Life") && connectMain.indexOf("The main accounts") >= 0);
for (const s of ZFL) check(`Connect page links ${s}`, html["connect.html"].includes(s));
check("Zoe Family Life X absence is stated", /No X account for Zoe Family Life/.test(html["connect.html"]));

check("Subscribe intro copy is present",
  publishedHtml.includes("Subscribe to receive encouragement, updates, and helpful resources from Zoe Life."));
check("Subscribe consent copy is present",
  publishedHtml.includes("I agree to receive emails and other communications, including marketing, from Zoe Life."));

/* ------------------------------------------------------------ styling -- */
group("Visual direction");

check("No dark page theme", !/--ink:\s*#(?:0|1|2)[0-9a-f]{5}\b/i.test(css) || /--cream:\s*#F/i.test(css));
check("Background is a light cream", /--cream:\s*#F7F4EE/i.test(css) && /--paper:\s*#F7F4EE/i.test(css));
check("Palette includes grey, tan, and peach", /--grey:\s*#E6E1DA/i.test(css) && /--tan:\s*#E4D3B8/i.test(css) && /--peach:\s*#F3D5B8/i.test(css));
check("theme-color is Instagram grey", PAGES.every((p) => /<meta name="theme-color" content="#E6E1DA">/.test(html[p])));
check("Reduced motion is respected", /prefers-reduced-motion:\s*reduce/.test(css));
check("Focus is visible", /:focus-visible/.test(css) && /outline:\s*3px/.test(css));
check("Type pairing is Fraunces and Outfit", /font-family: Fraunces/.test(css) && /font-family: Outfit/.test(css));
check("No Inter or Roboto as the only sans", !/--sans:\s*Inter|--sans:\s*Roboto/.test(css));

const scriptSrcs = [...publishedHtml.matchAll(/<script[^>]*src="([^"]+)"/g)].map((m) => m[1]);
check(
  "No WebGL or heavy animation libraries loaded",
  !scriptSrcs.some((s) => /three|gsap|lenis|scrolltrigger/i.test(s)),
  scriptSrcs.join(", ")
);
check(
  "Only local site scripts are loaded",
  scriptSrcs.every((s) => s === "js/main.js" || s === "js/config.js"),
  scriptSrcs.join(", ")
);
check("Tap targets are at least 44px", /min-height:\s*4[48]px/.test(css));

for (const p of PAGES) {
  const text = visibleText(html[p]);
  const dashes = text.match(/[—–]/g) || [];
  check(`${p} has no em or en dashes in visible copy`, dashes.length === 0, `${dashes.length} found`);
}

/* ----------------------------------------------------------- contrast -- */
group("Colour contrast (WCAG 2.1)");

function srgb(c) {
  c /= 255;
  return c <= 0.03928 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
}
function luminance(hex) {
  const n = parseInt(hex.replace("#", ""), 16);
  const r = (n >> 16) & 255, g = (n >> 8) & 255, b = n & 255;
  return 0.2126 * srgb(r) + 0.7152 * srgb(g) + 0.0722 * srgb(b);
}
function ratio(a, b) {
  const [x, y] = [luminance(a), luminance(b)].sort((m, n) => n - m);
  return (x + 0.05) / (y + 0.05);
}

const token = (name) => {
  const m = css.match(new RegExp(`--${name}:\\s*(#[0-9A-Fa-f]{6})`));
  if (!m) throw new Error(`token --${name} not found in css/style.css`);
  return m[1];
};

const T = {
  cream: token("cream"),
  creamDeep: token("cream-deep"),
  grey: token("grey"),
  tan: token("tan"),
  peach: token("peach"),
  sun: token("sun"),
  white: token("white"),
  goldDeep: token("gold-deep"),
  sageDeep: token("sage-deep"),
  terracotta: token("terracotta"),
  ink: token("ink"),
  inkSoft: token("ink-soft"),
};

const PAIRS = [
  ["body text on cream", T.ink, T.cream, 4.5],
  ["body text on cream-deep", T.ink, T.creamDeep, 4.5],
  ["body text on grey", T.ink, T.grey, 4.5],
  ["body text on tan", T.ink, T.tan, 4.5],
  ["body text on peach", T.ink, T.peach, 4.5],
  ["body text on sun", T.ink, T.sun, 4.5],
  ["muted text on cream", T.inkSoft, T.cream, 4.5],
  ["muted text on cream-deep", T.inkSoft, T.creamDeep, 4.5],
  ["muted text on grey", T.inkSoft, T.grey, 4.5],
  ["muted text on tan", T.inkSoft, T.tan, 4.5],
  ["muted text on peach", T.inkSoft, T.peach, 4.5],
  ["muted text on sun", T.inkSoft, T.sun, 4.5],
  ["muted text on white", T.inkSoft, T.white, 4.5],
  ["link sage-deep on cream", T.sageDeep, T.cream, 4.5],
  ["link sage-deep on white", T.sageDeep, T.white, 4.5],
  ["eyebrow gold-deep on cream", T.goldDeep, T.cream, 4.5],
  ["eyebrow gold-deep on cream-deep", T.goldDeep, T.creamDeep, 4.5],
  ["eyebrow gold-deep on grey", T.goldDeep, T.grey, 4.5],
  ["eyebrow gold-deep on tan", T.goldDeep, T.tan, 4.5],
  ["eyebrow gold-deep on peach", T.goldDeep, T.peach, 4.5],
  ["eyebrow gold-deep on sun", T.goldDeep, T.sun, 4.5],
  ["eyebrow gold-deep on white", T.goldDeep, T.white, 4.5],
  ["tagline terracotta on cream", T.terracotta, T.cream, 4.5],
  ["tagline terracotta on cream-deep", T.terracotta, T.creamDeep, 4.5],
  ["pending badge text on peach tint", T.terracotta, "#FBEDE2", 4.5],
  ["error text on cream", "#8A3E17", T.cream, 4.5],
  ["white text on sage-deep band", "#F6F4EE", T.sageDeep, 4.5],
  ["primary button label", T.white, T.sageDeep, 4.5],
  ["focus ring on cream", T.sageDeep, T.cream, 3],
];

for (const [name, fg, bg, min] of PAIRS) {
  const r = ratio(fg, bg);
  check(`${name} (${r.toFixed(2)}:1, needs ${min}:1)`, r >= min);
}

/* ---------------------------------------------------------------- done -- */
console.log(`\n${"=".repeat(60)}`);
if (failures.length) {
  console.log(`FAILED: ${failures.length} check(s), ${pass} passed\n`);
  failures.forEach((f) => console.log(`  - ${f}`));
  process.exit(1);
}
console.log(`PASSED: all ${pass} checks`);
